hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/cdc/TestCDCDataFrameSuite.scala

```diff
@@ -0,0 +1,238 @@
+
+  var spark: SparkSession = _
+
+  val commonOpts = Map(
+    HoodieTableConfig.CDC_ENABLED.key -> "true",
+    "hoodie.insert.shuffle.parallelism" -> "4",
+    "hoodie.upsert.shuffle.parallelism" -> "4",
+    "hoodie.bulkinsert.shuffle.parallelism" -> "2",
+    "hoodie.delete.shuffle.parallelism" -> "1",
+    RECORDKEY_FIELD.key -> "_row_key",
+    PRECOMBINE_FIELD.key -> "timestamp",
+    HoodieWriteConfig.TBL_NAME.key -> "hoodie_test",
+    HoodieMetadataConfig.COMPACT_NUM_DELTA_COMMITS.key -> "1",
+    HoodieCleanConfig.AUTO_CLEAN.key -> "false"
+  )
+
+  @BeforeEach override def setUp(): Unit = {
+    setTableName("hoodie_test")
+    initPath()
+    initSparkContexts()
+    spark = sqlContext.sparkSession
+    initTestDataGenerator()
+    initFileSystem()
+  }
+
+  @AfterEach override def tearDown(): Unit = {
+    cleanupSparkContexts()
+    cleanupTestDataGenerator()
+    cleanupFileSystem()
+  }
+
+  @ParameterizedTest
+  @CsvSource(Array("cdc_op_key", "cdc_data_before", "cdc_data_before_after"))
+  def testCOWDataSourceWrite(cdcSupplementalLoggingMode: String): Unit = {
+    val options = commonOpts ++ Map(
+      HoodieTableConfig.CDC_SUPPLEMENTAL_LOGGING_MODE.key -> cdcSupplementalLoggingMode
+    )
+
+    // Insert Operation
+    val records1 = recordsToStrings(dataGen.generateInserts("000", 100)).toList
+    val inputDF1 = spark.read.json(spark.sparkContext.parallelize(records1, 2))
+    inputDF1.write.format("org.apache.hudi")
+      .options(options)
+      .mode(SaveMode.Overwrite)
+      .save(basePath)
+
+    // init meta client
+    metaClient = HoodieTableMetaClient.builder()
+      .setBasePath(basePath)
+      .setConf(spark.sessionState.newHadoopConf)
+      .build()
+    val instant1 = metaClient.reloadActiveTimeline.lastInstant().get()
+    assertEquals(spark.read.format("hudi").load(basePath).count(), 100)
+    // all the data is new-coming, it will write out cdc log files.
+    assertFalse(hasCDCLogFile(instant1))
+
+    val schemaResolver = new TableSchemaResolver(metaClient)
+    val dataSchema = schemaResolver.getTableAvroSchema(false)
+    val cdcSchema = HoodieCDCUtils.schemaBySupplementalLoggingMode(
+      HoodieCDCSupplementalLoggingMode.parse(cdcSupplementalLoggingMode), dataSchema)
+
+    // Upsert Operation
+    val records2 = recordsToStrings(dataGen.generateUniqueUpdates("001", 50)).toList
+    val inputDF2 = spark.read.json(spark.sparkContext.parallelize(records2, 2))
+    inputDF2.write.format("org.apache.hudi")
+      .options(options)
+      .mode(SaveMode.Append)
+      .save(basePath)
+    val instant2 = metaClient.reloadActiveTimeline.lastInstant().get()
+    assertEquals(spark.read.format("hudi").load(basePath).count(), 100)
+
+    // part of data are updated, it will write out cdc log files
+    assertTrue(hasCDCLogFile(instant2))
+    val cdcData2 = getCDCLogFIle(instant2).flatMap(readCDCLogFile(_, cdcSchema))
+    assertEquals(cdcData2.size, 50)
```

Comment:
Let's make sure we add test (either modifying this one or adding a new one) that will be asserting log-files contents as well

Reply:
done